### DISCUSSION TITLE
Fine-grained pathname generation

### DISCUSSION THREAD

**Discussion by smastrom-dolly** [withastro/roadmap#235] [Proposal]
Being a fan of SSG, yesterday I decided to nerd a bit with Astro. It's fast, simple and super-flexible. Congrats to the team, seriously.

However, I noticed that file-based page generation with square brackets has the same limits as NextJS.

The most basic (useless) example can be the following one, I have two pages where slugs come from a CMS:
- /about
- /contact

By using square brackets file naming, I can't use different .astro files for each page unless I statically name each file.

A real world example can be a multilanguage site:
- `/en/blog/article-name` | `/it/articoli/nome-articolo`
- `/it/recipes/spaghetti-carbonara` | `/it/ricette/spaghetti-carbonara`

The only way for `article-name` and `spaghetti-carbonara` to have different layouts is to have the parent folder statically named so I can create two .astro files in two different folders:
- `[locale]/blog/[slug]` | `[locale]/blog/[slug]`
- `[locale]/recipes/[slug]` | `[locale]/recipes/[slug]`

Sometimes, it happens that pages have the same pathname structure but different templates and slugs are not going to be managed by the developer itself. 

For example, Gatsby and Nuxt have such features to give more control over the pathname.

May Astro do something similar?

**Comment by L1lith** [comment]
> Sometimes, it happens that pages have the same pathname structure but different templates and slugs are not going to be managed by the developer itself.
> 
> For example, Gatsby and Nuxt have such features to give more control over the pathname.
> 
> May Astro do something similar?

I believe [this](https://docs.astro.build/en/core-concepts/routing/#dynamic-routes) is the closest thing there is atm.

I was running into this issue when I wanted to do internationalized routing. I agree it would be cool to automatically map our page
```
/pages/apple.jsx
```
to a bunch of localized urls
```
/en/apple
/es/manzana
/fr/pomme
...etc
```

Perhaps there could be some kind of directory specific component which could control the routing. Something like `_routes.js` that could receive a child set of files and directories as a "routes" prop, and export some kind of routing structure?

```js
const languages = ["en", "es", "fr"]
export default function getRoutes(routes) {
    // Here our routes are generated, notice we assign the changing lang prop as well to each "virtual page"
    return languages.map(lang => routes.map(route => ({from: route, to: lang + "/" + route, lang }))).flat(Infinity)
}
```

Being able to change descendant routes would also enable us to do things like automatic localization of routes

**Comment by jasikpark** [comment]
You could probably do something like we do in our site for blog pages + blog index pages:

we put our blog markdown in `src/data/posts/*.mdx` and then use `getStaticPaths` to import them. We add `isPost: true` into the posts when we merge them w/ the `paginate` function output, and then render with:

```astro

{page.isPost ? (
  <BlogPost content={page}>
    <page.Content />
  </BlogPost>
) : (
  <BlogIndex meta={BlogMeta} page={page} />
)}
```

so we end up with different astro files used for the layouts, right?

full file: 

```astro
---
/** `src/pages/blog/[...slug].astro` */
import BlogIndex from "@layouts/BlogIndex.astro";
import BlogPost from "@layouts/BlogPost.astro";
import BlogMeta from "./meta.json";

interface MarkdownFrontmatter {
  title: string;
  hide?: boolean;
  description: string;
  slug: string;
  date: string;
}

export async function getStaticPaths({ paginate }) {
  const meta = await import("./meta.json");
  const basePath = "/blog/";
  const origin = Astro.site.origin;

  // get all posts that have a slug and date and aren't hidden, and sort by date.
  const allPages = (await Astro.glob<MarkdownFrontmatter>("../../data/posts/*.md"))
    .filter((post) => !!post.frontmatter.slug && !!post.frontmatter.date && !post.frontmatter.hide)
    .sort(
      (a, b) => new Date(b.frontmatter.date).valueOf() - new Date(a.frontmatter.date).valueOf()
    );

  // generate paths for the blog posts
  const postPaths = allPages.map(({ Content, frontmatter }) => {
    return {
      params: {
        page: frontmatter.slug,
      },
      props: {
        page: {
          ...frontmatter,
          Content,
          isPost: true,
        },
      },
    };
  });

  // generate paths for the blog index pages
  const paginationPaths = paginate(allPages, { pageSize: 10 });

  return [...paginationPaths, ...postPaths];
}

const { page } = Astro.props;
---

{page.isPost ? (
  <BlogPost content={page}>
    <page.Content />
  </BlogPost>
) : (
  <BlogIndex meta={BlogMeta} page={page} />
)}
```

P.S. this is pulled from a 1.0-beta site, so I'm mostly sure it works unchanged, but I haven't taken a look at the 1.0 version yet.

**Reply by jasikpark** [reply]
Though I suppose another caveat is that we don't do anything creative with the slugs today, like building them off of frontmatter values

**Reply by L1lith** [reply]
```js
  const allPages = (await Astro.glob<MarkdownFrontmatter>("../../data/posts/*.md"))
    .filter((post) => !!post.frontmatter.slug && !!post.frontmatter.date && !post.frontmatter.hide)
    .sort(
      (a, b) => new Date(b.frontmatter.date).valueOf() - new Date(a.frontmatter.date).valueOf()
    );
 ```
 To me it makes more sense as a pattern to have our post pages in the `/blog` directory directly, because after all those MD files do represent the blog posts. If I used this approach to translate an entire site that means I gotta have a separate dir for all my pages, which is why it might make sense to have some kind of "middleware" like approach that takes the existing routes, but maybe that's just me.

**Comment by smastrom-dolly** [comment]
I think would be great to have an alternative way of generating/adding paths. It could be done in a config file created for such purpose:

**astro-paths.js**
```js
import { AboutPage } from '@components/AboutPage';
import { ContactPage } from '@components/ContactPage';

export async function addStaticPaths() {
    const paths = [];
    
    // Logic to get pages data arrays, locales...
    
    aboutPages.forEach(({ lang, slug }) => {
         const page = {
             path: lang === 'en' ? `/${slug}` : `/${lang}/${slug}`,
             component: AboutPage,
             // any other data...
         };
         paths.push(page);
    });
    
    contactPages.forEach(({ lang, slug }) => {
         const page = {
             path: lang === 'en' ? `/${slug}` : `/${lang}/${slug}`,
             component: ContactPage,
             // any other data...
         };
         paths.push(page);
    });
    
    return paths
}
```

But I'm pretty sure there are better ways of doing that and that the team surely knows how to.
